README.md
# pbitmtopbm
pbitm bitmap converters for PBM, these might be of historic interest
to someone.

These programs are GPL'ed, see COPYING for license details.

These tools convert between the pbitm bitmap format from the prc-tools
and the standard PBM bitmap format.

There shouldn't be any problems (the code is dead simple) unless you have
some bizarre system or you lack an ANSI compiler.  You'll need the PBM
libraries and headers installed somewhere or nothing will work.

	Eric Howe
	mu@trends.net
	http://www.trends.net/~mu


ChangeLog
Version 1.4, 1998.12.18:
	- hacked up to use automake/autoconf
	- various minor reorganizations for autoconf


common.c
/*
 * common.c
 *	Common stuff for pbitmtopbm and pbmtopbitm.
 *
 * Copyright (C) 1997 Eric A. Howe
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *   Authors:  Eric A. Howe (mu@trends.net)
 */
#include <mine/rcs.h>
MU_ID("$Mu: pbitmtopbm/common.c 1.6 1998/12/18 19:47:18 $")

#include <mine/common.h>

char *
err(int e)
{
	char	*s;

	assert(e != 0);
	switch(e) {
	case E_BADROWLEN:	s = "Row with wrong length";		break;
	case E_BADPBM:		s = "Busted pbm file";			break;
	case E_TOOLONG:		s = "Too many columns for pbitm file";	break;
	default:		s = strerror(e);			break;
	}
	return s;
}

void
die(const char *fmt, ...)
{
	va_list	ap;

	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	exit(EXIT_FAILURE);
}


mine/rcs.h
/*
 * rcs.h
 *	RCS keyword imbedding macros for wlib.
 *
 * Copyright (C) 1996  Eric A. Howe
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *   Authors:  Eric A. Howe (mu@trends.net)
 */
#ifndef	PBITMTOPBM_RCS_H
#define	PBITMTOPBM_RCS_H

/*
 * A tricky way to get a string in a compiled file without polluting the
 * global name space.  The little mess will also keep 'gcc -Wall' quiet
 * (thus allowing you to use '-Werror' to be a bastard _and_ get your
 * version strings in the compiled file).  Note that we prefix a "@(#)"
 * on the version strings so that what(1) (from SCCS) can find the
 * version strings too.
 *
 * Use MU_ID() for C files and MU_HID() for headers
 */
#define	MU_IDBASE(s,x) \
		static char	Mu_iD_##s[]     = "@(#)" x; \
		static int	Mu_iD_trick_##s = sizeof(Mu_iD_trick_##s) \
						+ sizeof(Mu_iD_##s);
#if defined(NORCS)
#	define	MU_ID(x)
#else
#	define	MU_ID(x)	MU_IDBASE(SoUrCeFiLe, x)
#endif

/*
 * almost never want these
 */
#define	NOHRCS
#if defined(NOHRCS)
#	define	MU_HID(s, x)
#else
#	define	MU_HID(s, x)	MU_IDBASE(s, x)
#endif

MU_HID(wlib_rcs_h, "$Mu: pbitmtopbm/mine/rcs.h 1.1 1998/12/18 19:47:07 $")

/*
 * standard C++ things
 */
#if defined(__cplusplus)
#	define	CDECLS_BEGIN	extern "C" {
#	define	CDECLS_END	}
#else
#	define	CDECLS_BEGIN
#	define	CDECLS_END
#endif

#endif


pbitmtopbm.c
/*
 * pbitmtopbm.c
 *	Pilot bitmap to PBM converter.  This converter expects input
 *	in the same format as txt2bitm from prc-tools.
 *
 * Copyright (C) 1997 Eric A. Howe
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *   Authors:  Eric A. Howe (mu@trends.net)
 */
#include <mine/rcs.h>
MU_ID("$Mu: pbitmtopbm/pbitmtopbm.c 1.17 1998/12/18 20:13:50 $")

#include <mine/common.h>

typedef struct {
	bit	**bitmap;	/* the bits 		*/
	int	rows, cols;	/* size of the bitmap	*/
	int	plain;		/* plain pbm format?	*/
} SLOP;

static int
translate(bit *dst, char *src, int len)
{
	int i;

	for(i = 0; i < len; ++i, ++src, ++dst)
		*dst = (*src != '-' && *src != ' ');

	return 0;
}

static int
readit(SLOP *s, FILE *fp)
{
	char	buf[1024];
	bit	**bitmap;
	int	rows, cols, len, i;

	bitmap = NULL;
	rows   = 0;
	cols   = 0;
	while(fgets(&buf[0], sizeof(buf), fp) != NULL) {
		++rows;
		len = strlen(buf);
		if(buf[len - 1] == '\n')
			buf[--len] = '\0';
		if(cols == 0)
			cols = len;
		else if(len != cols)
			return E_BADROWLEN;
		if(len > MAXCOLS)
			return E_TOOLONG;

		/*
		 * Yes, the dumb slow stupid way.  I could get fancy and
		 * reduce the reallocing but that would just be a waste
		 * of time and effort (and yes, I know this will leak
		 * if the realloc fails).
		 */
		if((bitmap = realloc(bitmap, rows * sizeof(bit *))) == NULL
		|| (bitmap[rows - 1] = calloc(cols, sizeof(bit))) == NULL)
			return ENOMEM;
		if((i = translate(bitmap[rows - 1], buf, len)) != 0)
			return i;
	}

	s->rows   = rows;
	s->cols   = cols;
	s->bitmap = bitmap;

	return 0;
}

static int
writeit(SLOP *s, FILE *fp)
{
	int r;

	pbm_writepbminit(fp, s->cols, s->rows, s->plain);
	for(r = 0; r < s->rows; ++r)
		pbm_writepbmrow(fp, s->bitmap[r], s->cols, s->plain);
	return 0;
}

static int
version(const char *me)
{
	printf("%s %s\n", me, VERSION);
	return EXIT_SUCCESS;
}

#define	OPTS	"hvtb"
static char *explain[] = {
	"\tConvert a pbitm file to a pbm file.",
	"",
	"\t-h\tDisplay this usage message and exit.",
	"\t-v\tDisplay the version number and exit.",
	"\t-t\tForce the pbm file to be in text format.",
	"\t-b\tForce the pbm file to be in raw/binary format.",
	NULL
};
static int
usage(const char *me, int ret)
{
	char	**s;

	printf("%s [-hv] [-t|-b] [in [out]]\n", me);
	for(s = &explain[0]; *s != NULL; ++s)
		printf("%s\n", *s);
	return ret;
}

int
main(int argc, char **argv)
{
	char	*me;
	SLOP	s;
	int	i;
	FILE	*in, *out;

	memset((void *)&s, '\0', sizeof(s));
	if((me = strrchr(*argv, '/')) == NULL)
		me = *argv;
	else
		++me;
	pbm_init(&argc, argv);
	while((i = getopt(argc, argv, OPTS)) != EOF) {
		switch(i) {
		case 'h':	exit(usage(me, EXIT_SUCCESS));		break;
		case 'v':	exit(version(me));			break;
		case 't':	s.plain = 1;				break;
		case 'b':	s.plain = 0;				break;
		default:	exit(usage(me, EXIT_FAILURE));		break;
		}
	}
	in = argv[optind] == NULL ? stdin : fopen(argv[optind++], "r");
	if(in == NULL)
		die("Could not open input %s: %s\n", argv[optind-1],err(errno));
	out = argv[optind] == NULL ? stdout : fopen(argv[optind], "w");
	if(out == NULL)
		die("Could not open output %s: %s\n", argv[optind], err(errno));

	if((i = readit(&s, in)) != 0)
		die("Could not parse input: %s\n", err(i));
	if((i = writeit(&s, out)) != 0)
		die("Could not write output: %s\n", err(i));

	/*
	 * These checks are probably silly but I didn't open stdin or
	 * stdout so I'm not going to close them dammit!
	 */
	if(in != stdin)
		fclose(in);
	if(out != stdout)
		fclose(out);

	return EXIT_SUCCESS;
}


mine/common.h
/*
 * common.h
 *	Common stuff for pbitmtopbm and pbmtopbitm.
 *
 * Copyright (C) 1997 Eric A. Howe
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *   Authors:	Eric A. Howe (mu@trends.net)
 */
#ifndef	PBITMTOPBM_COMMON_H
#define	PBITMTOPBM_COMMON_H
MU_HID(pbitmtopbm_common_h, "$Mu: pbitmtopbm/mine/common.h 1.2 1998/12/18 19:49:55 $")

#include <assert.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>

#include <pbm.h>

/*
 * Max number of columns for a pbitm file.
 */
#define	MAXCOLS	163

/*
 * These tend to live in different places on different systems so I'm cheating.
 */
extern	int	optind;
extern	char	*optarg;
extern	int	getopt(int, char *const *, const char *);

/*
 * Extra error numbers.
 */
#define	E_BADROWLEN	-1
#define	E_BADPBM	-2
#define	E_TOOLONG	-3

CDECLS_BEGIN
extern	char	*err(int);
extern	void	die(const char *fmt, ...);
CDECLS_END

#endif


pbmtopbitm.c
/*
 * pbmtopbitm.c
 *	PBM to Pilot bitmap. This converter outputs the format the
 *	txt2bitm from prc-tools expects.
 *
 * Copyright (C) 1997 Eric A. Howe
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *   Authors:  Eric A. Howe (mu@trends.net)
 */
#include <mine/rcs.h>
MU_ID("$Mu: pbitmtopbm/pbmtopbitm.c 1.11 1998/12/18 19:47:18 $")

#include <mine/common.h>

typedef struct {
	bit	**bitmap;	/* the bitmap, unpacked	*/
	int	rows, cols;	/* bitmap size		*/
	char	one, zero;	/* on/off characters	*/
} SLOP;

/*
 * Valid off/zero characters.  Technically I'm not allowed to
 * use "is[a-z]+" as a name here (all such names are reserved for
 * the standard libraries) but I'm going to live dangerously and
 * openly flaunt the standard (obviously my next step will be
 * blowing up french-fry trucks on crowded down-town streets
 * or some other equally unsociable act).
 */
#define	iszerochar(c)	((c) == ' ' || (c) == '-')

/*
 * Default zero and one characters.
 */
#define	CZERO	'-'
#define	CONE	'#'

static int
readit(SLOP *s, FILE *fp)
{
	bit	**bitmap;
	int	rows, cols;

	if((bitmap = pbm_readpbm(fp, &cols, &rows)) == NULL)
		return E_BADPBM;
	s->bitmap = bitmap;
	s->rows   = rows;
	s->cols   = cols;

	return 0;
}

static int
writeit(SLOP *s, FILE *fp)
{
	int	r, c;
	char	*buf;

	if(s->cols > MAXCOLS)
		return E_TOOLONG;

	if((buf = calloc(s->cols + 2, 1)) == NULL)
		return ENOMEM;
	buf[s->cols] = '\n';
	for(r = 0; r < s->rows; ++r) {
		for(c = 0; c < s->cols; ++c)
			buf[c] = s->bitmap[r][c] == 0 ? s->zero : s->one;
		fputs(buf, fp);
	}
	free((void *)buf);

	return 0;
}

static int
version(const char *me)
{
	printf("%s %s\n", me, VERSION);
	return EXIT_SUCCESS;
}

#define	OPTS	"hvo:z:"
char *explain[] = {
	"\tConvert a pbm file to a pbitm file.",
	"",
	"\t-h    Display this usage message and exit.",
	"\t-v    Display the version number and exit.",
	"\t-o c  Specify the character for one (on) bits, 'c' should be",
	"\t      anything but space or '-'",
	"\t-z c  Specify the character for zero (off) bits, 'c' should be",
	"\t      a space or '-'.",
	NULL
};
static int
usage(const char *me, int ret)
{
	char	**s;

	printf("%s [-hv] [-o c] [-z c] [in [out]]\n", me);
	for(s = &explain[0]; *s != NULL; ++s)
		printf("%s\n", *s);
	return ret;
}

int
main(int argc, char **argv)
{
	char	*me;
	SLOP	s;
	int	i;
	FILE	*in, *out;

	memset((void *)&s, '\0', sizeof(s));
	if((me = strrchr(*argv, '/')) == NULL)
		me = *argv;
	else
		++me;
	s.one  = CONE;
	s.zero = CZERO;
	pbm_init(&argc, argv);
	while((i = getopt(argc, argv, OPTS)) != EOF) {
		switch(i) {
		case 'h':	exit(usage(me, EXIT_SUCCESS));		break;
		case 'v':	exit(version(me));			break;
		case 'o':	s.one  = *optarg;			break;
		case 'z':	s.zero = *optarg;			break;
		default:	exit(usage(me, EXIT_FAILURE));		break;
		}
	}
	if(iszerochar(s.one))
		die("Invalid on/one character: %c\n", s.one);
	if(!iszerochar(s.zero))
		die("Invalid off/zero character: %c\n", s.zero);

	in = argv[optind] == NULL ? stdin : fopen(argv[optind++], "r");
	if(in == NULL)
		die("Could not open input %s: %s\n", argv[optind-1],err(errno));
	out = argv[optind] == NULL ? stdout : fopen(argv[optind], "w");
	if(out == NULL)
		die("Could not open output %s: %s\n", argv[optind], err(errno));

	if((i = readit(&s, in)) != 0)
		die("Could not parse input: %s\n", err(i));
	if((i = writeit(&s, out)) != 0)
		die("Could not write output: %s\n", err(i));

	pbm_freearray(s.bitmap, s.rows);

	if(in != stdin)
		fclose(in);
	if(out != stdout)
		fclose(out);

	return EXIT_SUCCESS;
}
